src/semanticlayertools/linkage/cocitation.py

```diff
@@ -0,0 +1,171 @@
+            reflen = chunk[self.columnName].apply(
+                lambda x: True if type(x) == list and len(x) <= self.limitRefLength else False
+            )
+            data = chunk[reflen].copy()
+        else:
+            data = chunk.copy()
+        for idx, row in data.iterrows():
+            comb = combinations(row[self.columnName], 2)
+            for elem in list(comb):
+                res.append((elem))
+        return res
+
+    def calculateCoCitation(self, filepath):
+        """Run calculation for single input file.
+
+        Creates three files: Metadata-File with all components information,
+        Giant component network data in pajek format and full graph data in
+        edgelist format.
+
+        :param filepath: Path for input corous
+        :type filepath: str
+        :returns: A tuple of GC information: Number of nodes and percentage of total, Number of edges and percentage of total
+        :rtype: tuple
+        """
+        infilename = filepath.split(os.path.sep)[-1].split('.')[0]
+        starttime = time.time()
+        try:
+            data = pd.read_json(filepath, lines=True).dropna(subset=[self.columnName])
+            chunk_size = int(data.shape[0] / self.numberProc)
+            if chunk_size == 0:  # Deal with small data samples.
+                chunk_size = 1
+            chunks = np.array_split(data, chunk_size)
+            pool = multiprocessing.Pool(processes=self.numberProc)
+            cocitations = pool.map(self.getCombinations, chunks)
+            cocitCounts = Counter([x for y in cocitations for x in y])
+            sortCoCitCounts = [
+                (x[0][0], x[0][1], x[1]) for x in cocitCounts.most_common()
+            ]
+            tempG = ig.Graph.TupleList(sortCoCitCounts, weights=True, vertex_name_attr='id')
+            components = tempG.components()
+            sortedComponents = sorted(
+                [(x, len(x), len(x) * 100 / len(tempG.vs)) for x in components], key=lambda x: x[1], reverse=True
+            )
+            with open(os.path.join(self.outpath, infilename + '_graphMetadata.txt'), 'w') as outfile:
+                outfile.write(f'Graph derived from {filepath}\nSummary:\n')
+                outfile.write(tempG.summary() + '\n\nComponents (ordered by size):\n\n')
+                for idx, elem in enumerate(sortedComponents):
+                    gcompTemp = tempG.vs.select(elem[0]).subgraph()
+                    outfile.write(
+                        f"{idx}:\n\t{elem[1]} nodes ({elem[2]:.3f}% of full graph)\n\t{len(gcompTemp.es)} edges ({len(gcompTemp.es)*100/len(tempG.es):.3f}% of full graph)\n\n"
+                    )
+                    if idx == 0:
+                        gcouttuple = (
+                            elem[1],
+                            elem[2],
+                            len(gcompTemp.es),
+                            len(gcompTemp.es) * 100 / len(tempG.es)
+                        )
```

Comment:
Cosmetics: This tuple is built twice, once inf line 123 and a second time here. Perhaps build it just once?

Reply:
i have to check if it is really the same tuple...